### PR TITLE
feat: add logic for token creation

### PR DESCRIPTION
**Pull request by LBokan**
### Description

Add logic for token creation, change DB tables

### Type of change

- [x] Feature (introduces new functionality)
- [ ] Bugfix
- [ ] Configuration
- [ ] Refactoring (doesn't change functionality, only the code)
- [ ] Style changes (doesn't affect functionality, only the look of the components)

### PR Checklist

- [x] Synced up with latest from the main branch/Code is linted
- [x] Description of the changes has been added (Screenshots/Video)

> If not explain here

- [x] Commit messages follow [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#summary)
- [x] Console has been checked for warnings

#### Screenshots/Video before change

Add screenshots of the changed components here.

#### Screenshots/Video after change

Add screenshots with the changes applied here.

### CODE REVIEW

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  private Claims extractAllClaims( String token ) {
+    return Jwts
+        .parserBuilder()
+        .setSigningKey( getSignInKey() )
+        .build()
+        .parseClaimsJws( token )
+        .getBody();
+  }
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java:37]: will throw `io.jsonwebtoken.ExpiredJwtException` if claim is not valid

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        .setIssuedAt( new Date( System.currentTimeMillis() ) )
+        .setExpiration( new Date( System.currentTimeMillis() + expirationMs ) )
+        .signWith( getSignInKey(), SignatureAlgorithm.HS256 )
+        .compact();
+  }
+
+  @Override
+  public boolean isTokenValid( String token, UserDetails userDetails ) {
+    final String email = extractEmail( token );
+    boolean isTokenExpired = extractAllClaims( token )
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java:53]: this this call will throw an exception

File: backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      if ( isTokenValid ) {
+        UsernamePasswordAuthenticationToken authToken = new UsernamePasswordAuthenticationToken(
+            userDetails,
+            null,
+            userDetails.getAuthorities()
+        );
+
+        authToken.setDetails(
+            new WebAuthenticationDetailsSource().buildDetails( request )
+        );
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java:63]: why creating auth manager if you are not going to use it

from auth manager doc
image

File: backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
+package com.exadel.carinsurance.config.jwt;
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java:1]: filter should not be located in config package.
Please use `config` package and it's subpackages only for classes related to project configuration such as bean definitions, utilities for computing data for bean initialization and etcetera. But not executable business code.

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java (hunk 1 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java:37]: Done

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java (hunk 2 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java:53]: Done

File: backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java (hunk 3 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java]: Done

File: backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java (hunk 4 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/config/jwt/JwtAuthFilter.java]: Done

File: backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package com.exadel.carinsurance.utils;
+
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+public class NoOpPasswordEncoder implements PasswordEncoder {
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java:6]: NoOp means `Non Operational` and used for implementations that do nothing. In your case you actually have a logic and actual validations in some of the cases. So please consider renaming it to something more appropriate as `TokenFilteringPasswordEncoder `

File: backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package com.exadel.carinsurance.utils;
+
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+public class NoOpPasswordEncoder implements PasswordEncoder {
+  @Override
+  public String encode( CharSequence rawPassword ) {
+    return rawPassword.toString();
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java:9]: as you use real password encoder under the hood, consider delegating to it's encoder as well

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    byte[] keyBytes = Decoders.BASE64.decode( secretKey );
+
+    return Keys.hmacShaKeyFor( keyBytes );
+  }
+
+  private Claims extractAllClaims( String token ) {
+    return Jwts
+        .parserBuilder()
+        .setSigningKey( getSignInKey() )
+        .build()
```
[the remark this answers is not in the document]
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/service/implementation/JwtService.java:34]: Done

File: backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  }
+
+  @Override
+  public boolean matches( CharSequence rawPassword,
+                          String encodedPassword ) {
+    if ("".contentEquals( rawPassword )) {
+      return true;
+    } else {
+      return new BCryptPasswordEncoder()
+          .matches( encode(rawPassword), encodedPassword );
```
[the remark this answers is not in the document]
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java:19]: Done

File: backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java (hunk 5 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java:6]: Done

File: backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java (hunk 6 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/utils/NoOpPasswordEncoder.java:9]: Done

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            new UsernameNotFoundException(
+                "A user with the email address does not exist"
+            ) );
+
+    authManager.authenticate(
+        new UsernamePasswordAuthenticationToken(
+            userEmail,
+            request.getPassword()
+        )
+    );
```
[the remark this answers is not in the document]
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java]: Done

**Review by rmaskalenka — APPROVED** [review]

File: backend/src/main/java/com/exadel/carinsurance/GlobalExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  public ResponseEntity usernameNotFoundException( UsernameNotFoundException ex ) {
+    return ResponseEntity
+        .status( HttpStatus.NOT_FOUND )
+        .body( ex.getMessage() );
+  }
+
+  @ExceptionHandler( BadCredentialsException.class )
+  public ResponseEntity badCredentialsException( BadCredentialsException ex ) {
+    return ResponseEntity
+        .status( HttpStatus.NOT_FOUND )
```
[the remark this answers is not in the document]
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/GlobalExceptionHandler.java:40]: Done

File: backend/src/main/java/com/exadel/carinsurance/utils/TokenFilteringPasswordEncoder.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
+package com.exadel.carinsurance.utils;
+
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+public class TokenFilteringPasswordEncoder implements PasswordEncoder {
+  private final PasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
```
[the remark this answers is not in the document]
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/utils/TokenFilteringPasswordEncoder.java]: Done

**Pull request merged**